File: RealEstateManager/app/src/main/java/com/openclassrooms/realestatemanager/data/database/RealEstateManagerDatabase.java
package com.openclassrooms.realestatemanager.data.database;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import android.content.Context;
import androidx.annotation.NonNull;


import com.openclassrooms.realestatemanager.data.dao.PropertiesDaoRoom;
import com.openclassrooms.realestatemanager.data.dao.entities.PropertyEntity;

import java.sql.Timestamp;
import java.util.Date;

@Database(entities = {PropertyEntity.class}, version = 1, exportSchema = false)
public abstract class RealEstateManagerDatabase extends RoomDatabase {

    // --- SINGLETON ---
    private static RealEstateManagerDatabase INSTANCE;

    // --- DAO ---
    public abstract PropertiesDaoRoom propertiesDaoRoom();

    public static void createInstanceTest(Context context) {
        INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                RealEstateManagerDatabase.class, new Timestamp(new Date().getTime()) + "ToDocTestAppDB.db")
                .allowMainThreadQueries()
                .addCallback(prepopulateDatabase())
                .build();
    }

    public static void createInstance(Context context) {
        if(INSTANCE == null) {
            INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                    RealEstateManagerDatabase.class, "realEstateManagerDB.db")
                    .addCallback(prepopulateDatabase())
                    .build();
        }
    }

    // --- INSTANCE ---
    public static RealEstateManagerDatabase getInstance() {
        return INSTANCE;
    }


    public static Callback prepopulateDatabase(){
        return new Callback() {

            @Override
            public void onCreate(@NonNull SupportSQLiteDatabase db) {
                super.onCreate(db);
            }
        };
    }
}

File: RealEstateManager/app/src/main/java/com/openclassrooms/realestatemanager/data/database/RealEstateManagerDatabase.kt
package com.openclassrooms.realestatemanager.data.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.sqlite.db.SupportSQLiteDatabase
import com.openclassrooms.realestatemanager.data.dao.PropertiesDaoRoom
import com.openclassrooms.realestatemanager.data.dao.converters.PictureEntityConverter
import com.openclassrooms.realestatemanager.data.dao.converters.StringConverter
import com.openclassrooms.realestatemanager.data.dao.entities.PictureEntity
import com.openclassrooms.realestatemanager.data.dao.entities.PropertyEntity
import java.sql.Timestamp
import java.util.*

@Database(entities = [
    PropertyEntity::class,
    PictureEntity::class
], version = 1, exportSchema = false)
@TypeConverters(
    PictureEntityConverter::class,
    StringConverter::class
)
abstract class RealEstateManagerDatabase : RoomDatabase() {
    // --- DAO ---
    abstract fun propertiesDaoRoom(): PropertiesDaoRoom?

    companion object {
        // --- INSTANCE ---
        // --- SINGLETON ---
        lateinit var instance: RealEstateManagerDatabase
            private set

        fun createInstanceTest(context: Context) {
            instance = Room.databaseBuilder(
                    context.applicationContext,
                    RealEstateManagerDatabase::class.java, Timestamp(Date().time).toString() + "ToDocTestAppDB.db")
                    .allowMainThreadQueries()
                    .addCallback(prepopulateDatabase())
                    .build()
        }

        fun createInstance(context: Context) {
            instance = Room.databaseBuilder(context.applicationContext,
                RealEstateManagerDatabase::class.java, "realEstateManagerDB.db")
                .addCallback(prepopulateDatabase())
                .build()
        }

        private fun prepopulateDatabase(): Callback {
            return object : Callback() {
                override fun onCreate(db: SupportSQLiteDatabase) {
                    super.onCreate(db)
                }
            }
        }
    }
}